### PR TITLE
Add internalRepoVisibility option to GitHub and GitLab connections

### PR DESCRIPTION
**Pull request by pjlast**
Adds an `internalRepoVisibility` option to GitHub and GitLab code host connections that can be either `"private"` or `"public"`.

This comes up enough to warrant a configurable option and not require the weird "add another code host connection" workaround.

TODO: Update docs

## Test plan

Tests added for different `internalRepoVisibilty` options.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
As you said, some docs will be good :)

**Review by kopancek — COMMENTED** [review]
This seems like a hack to me. It will break the permission reason column that I was just touching in 
#55394 .

Could we instead add internal as a column to the repo table? We could then easily modify authzQuery to do what we want - if the repo is internal, check if user has an external account linked to the same code host.

**Comment by pjlast** [comment]
@kopancek but that isn't what customers want. Some customers want internal repositories to be treated as public repositories. This means being visible even if the person hasn't connected an external account.

File: internal/repos/github.go [hunk 1]
```diff
@@ -389,6 +389,19 @@ func (s *GitHubSource) makeRepo(r *github.Repository) *types.Repo {
...
+	isPrivate := r.IsPrivate
+	if r.Visibility == "internal" {
+		switch s.config.InternalRepoVisibility {
+		case "public":
+			isPrivate = false
+		case "private":
+			isPrivate = true
+		default:
+			isPrivate = true
+		}
```
Comment (erzhtor) [inline internal/repos/github.go:402]: I guess, can be simplified as:
```suggestion
		isPrivate = s.config.InternalRepoVisibility !== 'public'
```

File: internal/repos/gitlab.go [hunk 2]
```diff
@@ -205,6 +206,19 @@ func (s GitLabSource) ExternalServices() types.ExternalServices {
...
+	isPrivate := proj.Visibility == "private"
+	if proj.Visibility == "internal" {
+		switch s.config.InternalRepoVisibility {
+		case "public":
+			isPrivate = false
+		case "private":
+			isPrivate = true
+		default:
+			isPrivate = true
+		}
```
Comment (erzhtor) [inline internal/repos/gitlab.go:219]: Same here

**Review by erzhtor — APPROVED** [review]

File: internal/repos/github.go (hunk 1 above)
Comment (pjlast) [reply internal/repos/github.go:402]: Be careful, your JavaScript is showing 😆 

But yeah definitely, thanks!

**Review by mrnugget — CHANGES_REQUESTED** [review]
This needs docs and a changelog if we want to land it.

**Pull request closed without merging**